### ISSUE TITLE
Mypy gives confusing/misleading error message

### ISSUE THREAD

**Issue by danpascu** [python/mypy#17531]
**Bug Report**

Mypy reports `error: Name "..." is not defined [name-defined]` when using an instance attribute that was annotated.

**To Reproduce**

```python
from inspect import Parameter


class ListElementDescriptor[T]:
    name: str
    item_type: type[T]

    @property
    def signature_parameter(self) -> Parameter:
        return Parameter(name=self.name, kind=Parameter.KEYWORD_ONLY, annotation=list[self.item_type])
```

**Expected Behavior**

Definitely not an error that `self.item_type` is not defined. I suspect mypy tries to to convey a different error, maybe related to using variables for type parameters in generics, though I disagree with that as well since list[] is not used here in a typing context to specify a type, but is used as a value.

**Actual Behavior**

```
typing-7.py:11: error: Name "self.item_type" is not defined  [name-defined]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: `mypy 1.12.0+dev.6a0657e5959ba1777c4d427f8f355d499035d145 (compiled: no)`
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files):
```
[tool.mypy]
enable_incomplete_feature = "NewGenericSyntax"
disable_bytearray_promotion = true
disable_memoryview_promotion = true
check_untyped_defs = true
warn_unreachable = true
warn_redundant_casts = true
warn_unused_ignores = true
```
- Python version used: 3.12
